key.go

```diff
@@ -0,0 +1,778 @@
+func (c Curve) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int(c))
+}
+
+// UnmarshalCBOR populates the KeyType form the provided CBOR value (must be
+// int or tstr).
+func (c *Curve) UnmarshalCBOR(data []byte) error {
+	var raw IntOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid curve value: %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := CurveFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*c = v
+	} else {
+		v := raw.Int()
+
+		if v < 1 || v > 7 {
+			return fmt.Errorf("unknown curve value %d", v)
+		}
+
+		*c = Curve(v)
+	}
+
+	return nil
+}
+
+// Key represents a COSE_Key structure, as defined by RFC8152.
+// Note: currently, this does NOT support RFC8230 (RSA algorithms).
+type Key struct {
+	keyStruct
+	// Curve is EC identifier -- taken form "COSE Elliptic Curves" IANA registry.
+	// Populated from keyStruct.RawKeyParam when key type is EC2 or OKP.
+	Curve Curve
+	// K is the key value. Populated from keyStruct.RawKeyParam when key
+	// type is Symmetric.
+	K []byte
+}
+
+// keyStruct embedded inside Key to enable two-tier unmarshalling, of KeyType
+// and parameter with label -1 (Curve or K, depending on key type).
+type keyStruct struct {
+	// Common parameters. These are independent of the key type. Only
+	// KeyType common parameter MUST be set.
+
+	// KeyType identifies the family of keys for this structure, and thus,
+	// which of the key-type-specific parameters need to be set.
+	KeyType KeyType `cbor:"1,keyasint" json:"kty"`
+	// KeyId is the identification value matched to the kid in the message.
+	KeyId []byte `cbor:"2,keyasint,omitempty" json:"kid,omitempty"`
+	// Algorithm is used to restrict the algorithm that is used with the
+	// key. If it is set, the application MUST verify that it matches the
+	// algorithm for which the Key is being used.
+	Algorithm Algorithm `cbor:"3,keyasint,omitempty" json:"alg,omitempty"`
+	// KeyOps can be set to restrict the set of operations that the Key is used for.
+	KeyOps []KeyOp `cbor:"4,keyasint,omitempty" json:"key_ops,omitempty"`
+	// BaseIv is the Base IV to be xor-ed with Partial IVs.
+	BaseIv []byte `cbor:"5,keyasint,omitempty" json:"base_iv,omitempty"`
+
+	// Key-type-specific parameters. Which of these need to be set, and
+	// what their values ought to be, is determined by the KeyType.
+
+	// RawKeyParam contains the raw CBOR encoded data for the label -1.
+	// Depending on the KeyType this is used to populate either Curve or K
+	// below.
+	RawKeyParam cbor.RawMessage `cbor:"-1,keyasint,omitempty"`
+
+	// EC2/OKP params
+
+	// X is the x-coordinate
+	X []byte `cbor:"-2,keyasint,omitempty"`
+	// Y is the y-coordinate (sign bits are not supported)
+	Y []byte `cbor:"-3,keyasint,omitempty"`
+	// D is the private key
+	D []byte `cbor:"-4,keyasint,omitempty"`
+}
+
+// CreateOkpKey returns a Key created using the provided Octet Key Pair data.
+func CreateOkpKey(curve Curve, x, d []byte) (*Key, error) {
```

Comment:
Using `New*` instead of `Create*` is more Goish. In this case, I would write `NewOkpKey` instead.